File: leetcode/Python/merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        pq, sol, cur = [], ListNode(), None
        for i,v in enumerate(lists):
            if v:
                heapq.heappush(pq,(v.val,i))
        cur=sol
        while pq:
            _,i = heapq.heappop(pq)
            cur.next=lists[i]
            cur, lists[i] =cur.next, lists[i].next
            if lists[i]:
                heapq.heappush(pq,(lists[i].val, i))
        return sol.next

File: leetcode/C++/merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        ListNode *sol=new ListNode(),*cur=sol;
        priority_queue<pair<int, int>> pq;
        int idx = 0;
        for(int i=0; i<lists.size(); ++i) if(lists[i]) pq.push({-lists[i]->val,i});
        while(!pq.empty()){
            idx = pq.top().second; pq.pop();
            cur->next=lists[idx]; cur=cur->next;
            if(lists[idx]->next) {
                lists[idx]=lists[idx]->next;
                pq.push({-lists[idx]->val,idx});
            }
        }
        return sol->next;
    }
};